Sprint5/step1

smuravyev opened the pull request:
Bugs fixed.

Action creators are used now. No plain actions.

To prevent doubling of sockets there is a check while processing web socket opening: if the socket is already connected and the URL is the same, it won't create new socket.  The socket will be closed only after all the components stopped using it.

The code is already formatted in my own style.

Will describe in russian:

К сожалению, в предыдущем PR были ошибки, связанные с экшном FEED_CONNECTING и его логикой. Из-за них возникали проблемы при открытии модальных окон: под ними исчезал список заказов.

В ходе предыдущей проверки (еще до ее окончания) баг был исправлен: статус bIsConnecting, который он ставил, нигде не использовался, поэтому этот статус убран, сам экшн тоже убран.

Теперь при открытии попапов все хорошо.

Если сокет уже открыт с тем же URL, он переиспользуется. То есть открытие нового сокета не происходит при повторном запросе на соединение с тем же сокетом. Внутри миддлвары реализован подсчет количества компонентов, запросивших подключение к этому сокету. Он закроется только тогда, когда таких компонентов останется ноль.

Файл App.test.tsx убрал.

gennady-bars reviewed (commented):
Уже почти все отлично, но:

* Вам не нужно закрывать `pull request`, пока работа не будет принята, чтобы видно было историю комментариев и изменений. Как только проектная работа будет решена, вот тогда нужно мерджить ветку в `main`.
* в истории заказов не отображаются заказы. Скрин http://joxi.ru/J2bzbWltgzXJD2 . 
* Иногда не проходит автологин. Какой-то баг

smuravyev commented:
Changes since last check:

1. Changed the conditions of the orders feed progress bar appearance: it waited for all the asynchronous actions to complete, and now it's waiting only for ingredients list and any valid data from the websocket.

2. Bug fixed: if the auth data was wrong, there was an infinite loading of the feed in the profile. Now it will try to re-login and if succeeded, show the feed, if not - jump to "/login/" route.

3. Bug fixed: if the socket connection is down, there will be only 3 attempts to reconnect with a 5 seconds interval. If none of them will succeeded, the unclosable error popup will be shown, and the customer has to refresh the page to proceed. The only choice. 

3. Found a layout bug at the profile orders page (the order plates sometimes did not width-grow to take all the available width). Fixed.

in Russian: 

1.  Поменял логику показа этого прогрессбара: он ждал окончания любых асинхронных действий, а сейчас ждет только загрузки списка ингредиентов и появления валидных данных в сокет-соединении. Замечу, что иногда бекенд "подтупливает", и пару-тройку секунд висит эта загрузка. 

2. Получилось воспроизвести бесконечную загрузку путем “порчи” access-токена (добавления в него произвольных символов). Если с таким токеном попробовать открыть ленту заказов, была бесконечная загрузка. Исправлено. Тут стоит заметить, что также иногда обновление токенов занимает несколько секунд.

3.  Получилось воспроизвести бесконечную загрузку путем отключения интернета посреди соединения (полагаю это будет в любой ситуации невалидного закрытия сокета). Теперь после трех неудачных попыток с интервалом 5 секунд будет показано сообщение об ошибке “Проверьте работу интернета“, а попытки преподключиться закончатся.

4. Ошибка в верстке исправлена.

К сожалению, не могу воспроизвести "иногда не происходит автологин": если в куках присутствует валидный токен, у меня авторизация проходила успешно, в т. ч. при "протухании" accessToken.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

p.s. оказалось, что проблема в заказах, сделанных мною 3-5 месяцев назад. В полях что-то поменялось, видимо. Я зарегал нового пользователя и создал заказы. Теперь все отлично отображается в истории.
